codex-rs/core/src/exec_policy.rs

```diff
@@ -84,58 +104,102 @@ pub(crate) async fn exec_policy_for(
-                } else {
-                    Some(ApprovalRequirement::NeedsApproval {
-                        reason: Some(reason),
-                    })
+pub(crate) async fn append_allow_prefix_rule_and_update(
+    codex_home: &Path,
+    current_policy: &Arc<RwLock<Policy>>,
+    prefix: &[String],
+) -> Result<(), ExecPolicyUpdateError> {
+    let policy_path = default_policy_path(codex_home);
+    blocking_append_allow_prefix_rule(&policy_path, prefix).map_err(|source| {
+        ExecPolicyUpdateError::AppendRule {
+            path: policy_path,
+            source,
+        }
+    })?;
+
+    current_policy
+        .write()
+        .await
+        .add_prefix_rule(prefix, Decision::Allow)?;
+
+    Ok(())
+}
+
+fn requirement_from_decision(
+    decision: Decision,
+    approval_policy: AskForApproval,
+) -> ApprovalRequirement {
+    match decision {
+        Decision::Forbidden => ApprovalRequirement::Forbidden {
+            reason: FORBIDDEN_REASON.to_string(),
+        },
+        Decision::Prompt => {
+            let reason = PROMPT_REASON.to_string();
+            if matches!(approval_policy, AskForApproval::Never) {
+                ApprovalRequirement::Forbidden { reason }
+            } else {
+                ApprovalRequirement::NeedsApproval {
+                    reason: Some(reason),
+                    allow_prefix: None,
                 }
             }
-            Decision::Allow => Some(ApprovalRequirement::Skip {
-                bypass_sandbox: true,
-            }),
+        }
+        Decision::Allow => ApprovalRequirement::Skip {
+            bypass_sandbox: true,
         },
-        Evaluation::NoMatch { .. } => None,
+    }
+}
+
+/// Return an allow-prefix option when a single plain command needs approval without
+/// any matching policy rule. We only surface the prefix opt-in when execpolicy did
+/// not already drive the decision (NoMatch) and when the command is a single
+/// unrolled command (multi-part scripts shouldn’t be whitelisted via prefix) and
+/// when execpolicy feature is enabled.
+fn allow_prefix_if_applicable(
+    commands: &[Vec<String>],
+    features: &Features,
+) -> Option<Vec<String>> {
+    if features.enabled(Feature::ExecPolicy) && commands.len() == 1 {
+        Some(commands[0].clone())
+    } else {
+        None
     }
 }
 
 pub(crate) async fn create_approval_requirement_for_command(
-    policy: &Policy,
+    exec_policy: &Arc<RwLock<Policy>>,
+    features: &Features,
     command: &[String],
     approval_policy: AskForApproval,
     sandbox_policy: &SandboxPolicy,
     sandbox_permissions: SandboxPermissions,
 ) -> ApprovalRequirement {
-    if let Some(requirement) = evaluate_with_policy(policy, command, approval_policy) {
-        return requirement;
-    }
+    let commands = parse_shell_lc_plain_commands(command).unwrap_or_else(|| vec![command.to_vec()]);
+    let policy = exec_policy.read().await;
+    let evaluation = policy.check_multiple(commands.iter());
```

Comment:
That `read()` call is holding a lock on the mutex, so you should`Drop` it as soon as you no longer need it (which releases the lock).

```suggestion
    let evaluation = {
        let policy = exec_policy.read().await;
        policy.check_multiple(commands.iter())
    };
```

also, a single line will drop the lock at the end of the statement, so this would also be OK:

```rust
let evaluation = exec_policy.read().await.check_multiple(commands.iter());
```